File: SWEA/D2/1954. 달팽이 숫자/달팽이 숫자.py
'''
- depth를 다루는 반복문
- x축을 다루는 반복문
- y축을 다루는 반복문
- 시작 좌표
- 방향을 바꾸는 변수
'''
t = int(input()) # 테스트 케이스
for test_case in range(1, t+1):
    n = int(input()) # 달팽이 크기
    snail = [[0]*n for i in range(n)] # 빈 달팽이 생성, 2차원 배열로 구현
    x, y = 0, -1 # 2차원 배열 시작 좌표
    trans = 1 # 방향 변수
    num = 1 # 할당될 값

    print("#%d" %test_case)
    while n > 0: # depth 반복문
        for i in range(n): # x 축 반복문
            y += trans
            snail[x][y] = num
            num += 1
        n -= 1 # 칸수 계산
        
        for i in range(n): # y 축 반복문
            x += trans
            snail[x][y] = num
            num += 1
        trans *= -1 # 방향 전환, 2줄 마다 변환

    for leng in range(len(snail)):
        print(*snail[leng])

File: SWEA/D2/1954. 달팽이 숫자/달팽이 숫자.java

import java.util.Scanner;
import java.io.FileInputStream;
import java.io.*;
import java.util.StringTokenizer;

class Solution {
    static BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static StringTokenizer st;
    public static void main(String[] args) throws IOException {
        //달팽이 길찾기
        //입력 1: T
            // testCase
        //입력 2: N
            // N * N 달팽이
            // 1 <= N <= 10
        //N * N 빈 배열 생성
            //달팽이 빈 공간 생성

        //concept
        //1, -1 이라는 방향 변수를 지정하여 좌표를 이동하게 함
        //좌표를 이동하며 1씩 늘어감
        //직진으로 진행할 변의 갯수 == (2*n - 1)
            // 1개를 먼저 진행시키고 n번의 반복문 실행
        int T = Integer.parseInt(bf.readLine());
        for (int tc = 1; tc <= T; tc++) {
            int N = Integer.parseInt(bf.readLine());
            int NPrint = N;
            int[][] snail = new int[N][N];
            int x = 0;
            int y = -1;
            int direction = 1;
            int num = 1; // 칸에 할당될 값

            while (N > 0){ // N번의 반복문 실행
                //가장 첫 변 먼저 실행
                for (int i = 0; i < N; i++) { // 변의 길이만큼 반복
                    y += direction;
                    snail[x][y] = num;
                    num++;
                }
                N -= 1;
                for (int i = 0; i < N; i++) {
                    x += direction;
                    snail[x][y] = num;
                    num++;
                }
                direction *= -1;
            }
            bw.write("#"+tc+"\n");
            for (int i = 0; i < NPrint; i++) {
                for (int j = 0; j < NPrint; j++) {
                    bw.write(snail[i][j] + " ");
                }
                bw.write("\n");
            }
            bw.flush();
        }
        bw.close();
    }
}
